core/src/main/java/org/infinispan/interceptors/impl/AsyncInvocationStage.java

```diff
@@ -61,20 +52,30 @@ public Object get() throws Throwable {
 
    @Override
    public boolean isDone() {
+      lockHandlers();
```

Comment:
I wonder if you should lock the handlers also with `toCompletableFuture()`, but given it's unsafe to assume `toCompletableFuture().isDone()` => `this.isDone()`, it's probably not necessary.